docs/programming-oak.md

```diff
@@ -328,9 +328,86 @@ passes this on to the Oak Runtime, which will then notify the running Nodes that
 termination has been requested (by returning `ERR_TERMINATED` on any current or
 future `oak.wait_on_channels()` invocations).
 
-## Channels and Handles
+## Nodes, Channels and Handles
+
+So far, we've only discussed writing a single Node for an Oak Application. This
+Node communicates with the outside world via a single channel, and the handle
+for the read half of this channel is acquired at start-of-day as the parameter
+to the `oak_main()` entrypoint. The other half of this single channel is a gRPC
+pseudo-Node, which passes on requests from external clients (and which is
+automatically created by the Oak Runtime at Application start-of-day).
+
+More sophisticated Applications are normally built from multiple interacting
+Nodes, for several reasons:
+
+- Dividing software into well-defined interacting components is a normal way to
+  reduce the overall complexity of software design.
+- Software that handles sensitive data, or which has additional privileges,
+  often separates out the parts that deal with this (the
+  ["principle of least privilege"](https://en.wikipedia.org/wiki/Principle_of_least_privilege)),
+  to reduce the blast radius if something goes wrong.
+- Information flow analysis can be more precise and fine-grained if components
+  are smaller and the interactions between them are constrained.
+
+The first step in building a multi-Node Application is to write the code for all
+of the Nodes; the `ApplicationConfiguration` needs to include the configuration
+and code for any Node that might get run as part of the Application. New Node
+types cannot be added after the application starts; any Node that the
+Application might need has to be included in the original configuration.
+
+As before, each Node must include an `oak_main(u64) -> u32` entrypoint, but for
+an internal Node it's entirely up to the Application developer as to what
+channel handle gets passed to this entrypoint, and as to what messages are sent
+down that channel. The application may choose to use protobuf-encoded messages
+(as gRPC does) for its internal communications, or something else entirely (e.g.
+the [serde crate](https://crates.io/crates/serde)).
+
+Regardless of how the application communicates with the new Node, the typical
+pattern is to:
+
+- Create a new channel with the
+  [`channel_create`](https://project-oak.github.io/oak/sdk/oak/fn.channel_create.html)
+  host function, receiving local handles for both halves of the channel.
+- Create a new Node instance with the
+  [`node_create`](https://project-oak.github.io/oak/sdk/oak/fn.node_create.html)
+  host function, passing in the handle for the read half of the new channel.
+- Afterwards, close the local handle for the read half, as it is no longer
+  needed.
+
+For example, the example Chat application creates a Node for each chat room and
```

Comment:
Link to https://github.com/project-oak/oak/tree/master/examples/chat

Reply:
Done.